File: backend/auth/google.ts
import { Router } from "express";
import type { Request, Response } from "express";
import prisma from "../../lib/prisma.js";
import { createSession, FRONTEND_URL } from "./helpers.js";

const router = Router();

const GOOGLE_CLIENT_ID = process.env.GOOGLE_CLIENT_ID;
const GOOGLE_CLIENT_SECRET = process.env.GOOGLE_CLIENT_SECRET;
const APP_URL = process.env.APP_URL || "http://localhost:3000";
const CALLBACK_URL = `${APP_URL}/auth/google/callback`;

interface GoogleTokenResponse {
    access_token: string;
    id_token: string;
    token_type: string;
    expires_in: number;
    refresh_token?: string;
}

interface GoogleUserInfo {
    sub: string;
    email: string;
    email_verified: boolean;
    name: string;
    picture?: string;
}

if (GOOGLE_CLIENT_ID && GOOGLE_CLIENT_SECRET) {
    // Step 1: Redirect to Google consent screen
    router.get("/google", (_req: Request, res: Response) => {
        const params = new URLSearchParams({
            client_id: GOOGLE_CLIENT_ID,
            redirect_uri: CALLBACK_URL,
            response_type: "code",
            scope: "openid email profile",
            access_type: "offline",
            prompt: "consent",
        });

        res.redirect(`https://accounts.google.com/o/oauth2/v2/auth?${params.toString()}`);
    });

    // Step 2: Handle callback — exchange code for tokens, fetch profile, upsert user
    router.get("/google/callback", async (req: Request, res: Response) => {
        try {
            const code = req.query.code as string | undefined;
            const error = req.query.error as string | undefined;

            if (error || !code) {
                res.redirect(`${FRONTEND_URL}/error?msg=Google+login+cancelled`);
                return;
            }

            // Exchange authorization code for tokens
            const tokenRes = await fetch("https://oauth2.googleapis.com/token", {
                method: "POST",
                headers: { "Content-Type": "application/x-www-form-urlencoded" },
                body: new URLSearchParams({
                    code,
                    client_id: GOOGLE_CLIENT_ID,
                    client_secret: GOOGLE_CLIENT_SECRET,
                    redirect_uri: CALLBACK_URL,
                    grant_type: "authorization_code",
                }),
            });

            if (!tokenRes.ok) {
                console.error("[google] Token exchange failed:", await tokenRes.text());
                res.redirect(`${FRONTEND_URL}/error?msg=Google+login+failed`);
                return;
            }

            const tokens = (await tokenRes.json()) as GoogleTokenResponse;

            // Fetch user profile from Google
            const userRes = await fetch("https://www.googleapis.com/oauth2/v3/userinfo", {
                headers: { Authorization: `Bearer ${tokens.access_token}` },
            });

            if (!userRes.ok) {
                console.error("[google] Failed to fetch user info:", await userRes.text());
                res.redirect(`${FRONTEND_URL}/error?msg=Google+login+failed`);
                return;
            }

            const profile = (await userRes.json()) as GoogleUserInfo;

            if (!profile.email) {
                res.redirect(`${FRONTEND_URL}/error?msg=No+email+from+Google`);
                return;
            }

            // Upsert user
            let user = await prisma.user.findUnique({ where: { email: profile.email } });
            if (!user) {
                user = await prisma.user.create({
                    data: {
                        email: profile.email,
                        name: profile.name ?? profile.email.split("@")[0] ?? "User",
                        avatar_url: profile.picture ?? null,
                        provider: "google",
                        is_verified: true,
                    },
                });
            } else {
                user = await prisma.user.update({
                    where: { email: profile.email },
                    data: {
                        name: user.name ?? profile.name,
                        avatar_url: user.avatar_url ?? profile.picture ?? null,
                        is_verified: true,
                    },
                });
            }

            // Upsert OAuthAccount
            await prisma.oAuthAccount.upsert({
                where: {
                    provider_provider_account_id: {
                        provider: "google",
                        provider_account_id: profile.sub,
                    },
                },
                create: {
                    userId: user.id,
                    provider: "google",
                    provider_account_id: profile.sub,
                    access_token: tokens.access_token,
                    refresh_token: tokens.refresh_token ?? null,
                },
                update: {
                    access_token: tokens.access_token,
                    refresh_token: tokens.refresh_token ?? null,
                },
            });

            // ✅ No token in URL — cookies handle everything
            const accessToken = await createSession(user.id, res, req)
            res.redirect(`${FRONTEND_URL}/dashboard`)
        } catch (err) {
            console.error("[google-callback]", err);
            res.redirect(`${FRONTEND_URL}/error?msg=Google+login+failed`);
        }
    });
} else {
    console.warn("[auth] Google OAuth not configured — missing GOOGLE_CLIENT_ID / GOOGLE_CLIENT_SECRET");
}

export default router;

File: backend/auth-service/app/routers/google.py
import httpx
from urllib.parse import urlencode

from fastapi import APIRouter, Request, Response
from fastapi.responses import RedirectResponse

from app.config import GOOGLE_CLIENT_ID, GOOGLE_CLIENT_SECRET, APP_URL, FRONTEND_URL
from app.database import get_pool
from app.routers._helpers import create_session

router = APIRouter()

CALLBACK_URL = f"{APP_URL}/auth/google/callback"


# ─── GET /google — Redirect to Google consent screen ────────────────

@router.get("/google")
async def google_redirect():
    if not GOOGLE_CLIENT_ID or not GOOGLE_CLIENT_SECRET:
        return {"error": "Google OAuth not configured"}

    params = urlencode({
        "client_id": GOOGLE_CLIENT_ID,
        "redirect_uri": CALLBACK_URL,
        "response_type": "code",
        "scope": "openid email profile",
        "access_type": "offline",
        "prompt": "consent",
    })
    return RedirectResponse(f"https://accounts.google.com/o/oauth2/v2/auth?{params}")


# ─── GET /google/callback — Exchange code, upsert user ──────────────

@router.get("/google/callback")
async def google_callback(request: Request, response: Response, code: str | None = None, error: str | None = None):
    try:
        if error or not code:
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=Google+login+cancelled")

        # Exchange authorization code for tokens
        async with httpx.AsyncClient() as client:
            token_res = await client.post(
                "https://oauth2.googleapis.com/token",
                data={
                    "code": code,
                    "client_id": GOOGLE_CLIENT_ID,
                    "client_secret": GOOGLE_CLIENT_SECRET,
                    "redirect_uri": CALLBACK_URL,
                    "grant_type": "authorization_code",
                },
                headers={"Content-Type": "application/x-www-form-urlencoded"},
            )

        if token_res.status_code != 200:
            print(f"[google] Token exchange failed: {token_res.text}")
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=Google+login+failed")

        tokens = token_res.json()

        # Fetch user profile from Google
        async with httpx.AsyncClient() as client:
            user_res = await client.get(
                "https://www.googleapis.com/oauth2/v3/userinfo",
                headers={"Authorization": f"Bearer {tokens['access_token']}"},
            )

        if user_res.status_code != 200:
            print(f"[google] Failed to fetch user info: {user_res.text}")
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=Google+login+failed")

        profile = user_res.json()

        if not profile.get("email"):
            return RedirectResponse(f"{FRONTEND_URL}/error?msg=No+email+from+Google")

        pool = get_pool()

        # Upsert user
        user = await pool.fetchrow("SELECT * FROM users WHERE email = $1", profile["email"])

        if not user:
            user = await pool.fetchrow(
                """
                INSERT INTO users (email, name, avatar_url, provider, is_verified)
                VALUES ($1, $2, $3, 'google', TRUE)
                RETURNING *
                """,
                profile["email"],
                profile.get("name") or profile["email"].split("@")[0] or "User",
                profile.get("picture"),
            )
        else:
            user = await pool.fetchrow(
                """
                UPDATE users SET
                    name = COALESCE(name, $2),
                    avatar_url = COALESCE(avatar_url, $3),
                    is_verified = TRUE
                WHERE email = $1
                RETURNING *
                """,
                profile["email"],
                profile.get("name"),
                profile.get("picture"),
            )

        # Upsert OAuthAccount
        await pool.execute(
            """
            INSERT INTO oauth_accounts (user_id, provider, provider_account_id, access_token, refresh_token)
            VALUES ($1, 'google', $2, $3, $4)
            ON CONFLICT (provider, provider_account_id) DO UPDATE SET
                access_token = EXCLUDED.access_token,
                refresh_token = EXCLUDED.refresh_token
            """,
            str(user["id"]),
            profile["sub"],
            tokens["access_token"],
            tokens.get("refresh_token"),
        )

        # Create session (cookies handle everything)
        await create_session(str(user["id"]), response, request)
        return RedirectResponse(f"{FRONTEND_URL}/dashboard")

    except Exception as e:
        print(f"[google-callback] {e}")
        return RedirectResponse(f"{FRONTEND_URL}/error?msg=Google+login+failed")
